Unittest standered11

mgaurav99 opened the pull request:
## What?
- Modified the existing test function for the 'cold' scenario.
- Added a new test function for the 'warm' scenario.
- Included additional comments for better code clarity

mgaurav99 committed:
One basic test function create.

mgaurav99 committed:
Added a unit test function for warm.
Improved existing function for cold.

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/WeatherReportService.cs:
```diff
@@ -0,0 +1,51 @@
+﻿using WeatherxUnit.Interfaces;
+
+public class WeatherReportService : IWeatherService
```
Interface should be named as IWeatherReportService

mgaurav99 replied:
done

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/IWeatherService.cs:
```diff
@@ -0,0 +1,22 @@
+    /// IWeatherService
+    /// Fetching weather report/data
+    /// </summary>
+    public interface IWeatherService
```
Interface should be named as IWeatherReportService

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/TemperatureService.cs:
```diff
@@ -0,0 +1,29 @@
+    ///  TemperatureService
+    ///  Calculates temperature
+    /// </summary>
+    public class TemperatureService : ITemperature
```
Inteface should be named as ITemperatureService

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/ITemperature.cs:
```diff
@@ -0,0 +1,21 @@
+    /// ITemperature
+    /// Calculates temperature
+    /// </summary>
+    public interface ITemperature
```
Inteface should be named as ITemperatureService

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/WeatherReportService.cs:
```diff
@@ -0,0 +1,51 @@
+
+    public string GetClimate(string city)
+    {
+
```
remove extra returns

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/WeatherReportService.cs:
```diff
@@ -0,0 +1,51 @@
+        _logger = logger;
+        _temperature = temperature;
+    }
+
```
watch out extra returns

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/WeatherReportService.cs:
```diff
@@ -0,0 +1,51 @@
+        }
+
+        int temp = _temperature.GetTemperature();  //mock for classes set up for function.
+
```
remove extra return

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/WeatherReportService.cs:
```diff
@@ -0,0 +1,51 @@
+        {
+            climate = "cold";
+
+
```
remove extra return

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/WeatherReportService.cs:
```diff
@@ -0,0 +1,51 @@
+        {
+            climate = "hot";
+        }
+
```
remove extra return

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/WeatherReportService.cs:
```diff
@@ -0,0 +1,51 @@
+            throw new ArgumentNullException("Provide city!");
+        }
+
+        int temp = _temperature.GetTemperature();  //mock for classes set up for function.
```
remove comment

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/WeatherReportService.cs:
```diff
@@ -0,0 +1,51 @@
+﻿using WeatherxUnit.Interfaces;
+
```
missing namespace

mgaurav99 replied:
could you elaborate bit more where i am missing namespace as i can see its included as 
 ﻿using WeatherxUnit.Interfaces;

vikchaturvedi replied:
Create an appropriate namespace and wrap the WeatherReportService class inside it. Currently, this class does not belong to any namespace.

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/ConsoleLogger.cs:
```diff
@@ -0,0 +1,48 @@
+    public class ConsoleLogger : ILogger
+    {
+        /// <summary>
+        /// Loggin errors 
```
spelling  mistake

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/ConsoleLogger.cs:
```diff
@@ -0,0 +1,48 @@
+        /// Logging warnings
+        /// </summary>
+        /// <param name="message"></param>
+
```
remove extra return

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/ConsoleLogger.cs:
```diff
@@ -0,0 +1,48 @@
+        /// Logging information
+        /// </summary>
+        /// <param name="message"></param>
+
```
remove extra return

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/ConsoleLogger.cs:
```diff
@@ -0,0 +1,48 @@
+        }
+
+        /// <summary>
+        /// Logging warnings
```
spelling mistake

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/TemperatureService.cs:
```diff
@@ -0,0 +1,29 @@
+        /// <returns>int</returns>
+        public int GetTemperature()
+        {
+
```
remove extra returns

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/TemperatureService.cs:
```diff
@@ -0,0 +1,29 @@
+    /// </summary>
+    public class TemperatureService : ITemperature
+    {
+
```
remove extra return

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/TemperatureService.cs:
```diff
@@ -0,0 +1,29 @@
+
+            Random random = new Random();
+            return  random.Next(10, 50);
+
```
remove extra return

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/TemperatureService.cs:
```diff
@@ -0,0 +1,29 @@
+    {
+
+        /// <summary>
+        ///  Getting  random temperature 
```
Should be Get instead of Getting

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/WeatherReportService.cs:
```diff
@@ -0,0 +1,51 @@
+    {
+
+
+        _logger.LogInfo($"This function called for city:{city} at {DateTime.Now} ");
```
Implement a DateTimeService to get current date time.
You can then mock and setup IDateTimeService interface with static date time in your test cases.

mgaurav99 replied:
Implemented the desired service and mocked to set up static date time

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,70 @@
+            string msg = _weatherService.GetClimate(city);
+
+            //Assert
+            _mockLogger.Verify(x => x.LogInfo(It.Is<string>(msg => msg.Contains($"This function called for city:{city}"))));
```
compare exact message and Implement a DateTimeService to get current date time.
You can then mock and setup IDateTimeService interface with static date time in your test cases.

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,70 @@
+        /// Testing for cold condition
+        /// </summary>
+        [Fact]
+        public void GetClimate_LocationIsMumbai_Returnscold() /// check name convention *
```
1. C should be capital in ReturnsCold.
2. Remove comment

Suggestion:
Naming convention for a test case looks more appropriate with conditions
For example : MethodName_StateUnderTest_ExpectedBehavior
GetClimate_LessThan10_ReturnsCold

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,70 @@
+{
+    /// <summary>
+    /// WeatherReportServiceTests
+    /// Testing WeatherReportService..
```
1. no need of two dots ..
2. Description should be "WeatherReportService test cases"

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,70 @@
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        public  WeatherReportServiceTests()
```
extra space after access modifier

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,70 @@
+    /// </summary>
+    public class WeatherReportServiceTests
+    {
+
```
remove extra return

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,70 @@
+
+            //Assert
+            _mockLogger.Verify(x => x.LogInfo(It.Is<string>(msg => msg.Contains($"This function called for city:{city}"))));
+            //_mockLogger.Verify(x => x.LogInfo($"This function called for city:{city} at {DateTime.Now} "));
```
remove commented line

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,70 @@
+        }
+
+        /// <summary>
+        /// Testing for warm condition
```
Test case for warm condition

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,70 @@
+        /// Testing for warm condition
+        /// </summary>
+        [Fact]
+        public void GetClimate_LocationIsDelhi_Returnswarm()
```
W should be capital in ReturnsWarm.

Suggestion:
Naming convention for a test case looks more appropriate with conditions
For example : MethodName_StateUnderTest_ExpectedBehavior
GetClimate_GreaterThan10AndLessThan30_ReturnsWarm

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,70 @@
+            _mockLogger.Verify(x => x.LogInfo(It.Is<string>(msg => msg.Contains($"This function called for city:{city}"))));
+            _mockTemperature.Verify(x => x.GetTemperature());
+            Assert.Equal(expectedValue, message);
+
```
extra return

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,70 @@
+            string message = _weatherService.GetClimate(city);
+
+            //Assert
+            _mockLogger.Verify(x => x.LogInfo(It.Is<string>(msg => msg.Contains($"This function called for city:{city}"))));
```
compare exact message and Implement a DateTimeService to get current date time.
You can then mock and setup IDateTimeService interface with static date time in your test cases.

vikchaturvedi reviewed (changes requested):
Requested few suggestions

mgaurav99 committed:
Added few more test cases.
Added service for datetime.

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/IDateTimeService.cs:
```diff
@@ -0,0 +1,21 @@
+        /// Returns the current date and time in a string format
+        /// </summary>
+        /// <returns></returns>
+        string GetCurrentDateTime();
```
This should return DateTime object

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/IDateTimeService.cs:
```diff
@@ -0,0 +1,21 @@
+        /// <summary>
+        /// Returns the current date and time in a string format
+        /// </summary>
+        /// <returns></returns>
```
provide return type in documentation

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/IWeatherService.cs:
```diff
@@ -0,0 +1,22 @@
+        /// Fetching climate on basis of city provided
+        /// </summary>
+        /// <param name="city"></param>
+        /// <returns></returns>
```
provide return type in documentation

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/IWeatherService.cs:
```diff
@@ -0,0 +1,22 @@
+        /// <summary>
+        /// Fetching climate on basis of city provided
+        /// </summary>
+        /// <param name="city"></param>
```
provide parameter description in documentation

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/ITemperature.cs:
```diff
@@ -0,0 +1,21 @@
+    public interface ITemperatureService
+    {
+        /// <summary>
+        ///  Getting random temperature for now
```
this should be Get and not getting

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/ITemperature.cs:
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using System.Collections.Generic;
```
remove unused usings

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/IDateTimeService.cs:
```diff
@@ -0,0 +1,21 @@
+﻿using System;
```
remove unused usings

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/ILogger.cs:
```diff
@@ -0,0 +1,33 @@
+﻿using System;
```
remove unused usings

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/ILogger.cs:
```diff
@@ -0,0 +1,33 @@
+{
+    /// <summary>
+    /// ILogger
+    /// Logging purposes
```
Comment should be
Interface to define log methods

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/ITemperature.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    /// <summary>
+    /// ITemperature
+    /// Calculates temperature
```
Comment should be
Defines a contract for retrieving temperature data.

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/IWeatherService.cs:
```diff
@@ -0,0 +1,22 @@
+    public interface IWeatherReportService
+    {
+        /// <summary>
+        /// Fetching climate on basis of city provided
```
avoid using ing in your comments. Use this description.

Provides a contract for retrieving weather report data.

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/ILogger.cs:
```diff
@@ -0,0 +1,33 @@
+    public interface ILoggerService
+    {
+        /// <summary>
+        /// Logging info
```
avoid using ing

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/WeatherReportService.cs:
```diff
@@ -0,0 +1,64 @@
+    /// <summary>
+    /// Constructor
+    /// </summary>
+    /// <param name="logger"></param>
```
provide description to the parameters in documentation

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/TemperatureService.cs:
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+using System.Collections.Generic;
```
remove unused usings

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/IWeatherService.cs:
```diff
@@ -0,0 +1,22 @@
+﻿using System;
```
remove unused usings

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/DateTimeService.cs:
```diff
@@ -0,0 +1,26 @@
+﻿using System;
```
remove unused usings

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/ConsoleLoggerService.cs:
```diff
@@ -0,0 +1,48 @@
+﻿using System;
+using System.Collections.Generic;
```
remove unused usings

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/ConsoleLoggerService.cs:
```diff
@@ -0,0 +1,48 @@
+{
+    /// <summary>
+    /// ConsoleLogger
+    /// Logging related implementations
```
description could be more appropriate like
Provides logging functionality by writing messages to the console with appropriate formatting based on the log level.

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/ConsoleLoggerService.cs:
```diff
@@ -0,0 +1,48 @@
+        /// <summary>
+        /// Logging errors 
+        /// </summary>
+        /// <param name="message"></param>
```
give proper description to all parameters in this class documentation

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/DateTimeService.cs:
```diff
@@ -0,0 +1,26 @@
+        /// Returns the current date and time in a string format
+        /// </summary>
+        /// <returns></returns>
+        public string GetCurrentDateTime()
```
return DateTime type

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/DateTimeService.cs:
```diff
@@ -0,0 +1,26 @@
+        /// <returns></returns>
+        public string GetCurrentDateTime()
+        {
+            return DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss");
```
DateTime.Now can create inconsistency while communicating with different systems as:
DateTime.Now returns local time which may differ between environments (example dev vs server)

Better alternative: Use DateTime.UtcNow if:
1. You want consistency across systems or regions.
2. You store timestamps in databases for later conversion.
3. Server should always return UTC time and then client scripts can change it in local timezone.

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,159 @@
+namespace WeatherReportTest
+{
+    /// <summary>
+    /// WeatherReportServiceTests
```
provide an appropriate documentaion
Suggestion
/// <summary>
/// Contains unit tests for the <see cref="WeatherReportService"/> class,
/// validating its behavior under various temperature conditions and input scenarios.
/// </summary>

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,159 @@
+            //Arrange
+            string sampleDate = "2023-10-01 12:00:00";
+            _mockTemperatureService.Setup(x => x.GetTemperature()).Returns(9);
+            _mockDateTimeService.Setup(x => x.GetCurrentDateTime()).Returns(sampleDate);
```
mockDateTimeService.Setup is common to all test cases and can be defined one time inside constructor

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/ITemperature.cs:
```diff
@@ -0,0 +1,21 @@
+    /// ITemperature
+    /// Calculates temperature
+    /// </summary>
+    public interface ITemperatureService
```
rename Interface file name and keep it synced with Interface name

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Interfaces/ILogger.cs:
```diff
@@ -0,0 +1,33 @@
+    /// ILogger
+    /// Logging purposes
+    /// </summary>
+    public interface ILoggerService
```
rename Interface file name and keep it synced with Interface name

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/DateTimeService.cs:
```diff
@@ -0,0 +1,26 @@
+        /// <summary>
+        /// Returns the current date and time in a string format
+        /// </summary>
+        /// <returns></returns>
```
provide return type in documentation

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/TemperatureService.cs:
```diff
@@ -0,0 +1,26 @@
+    public class TemperatureService : ITemperatureService
+    {
+        /// <summary>
+        ///  Get  random temperature 
```
extra space in between of Get and random

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/ConsoleLoggerService.cs:
```diff
@@ -0,0 +1,48 @@
+    public class ConsoleLoggerService : ILoggerService
+    {
+        /// <summary>
+        /// Logging errors 
```
I would suggest avoid ing and use descriptions like
Log errors

vikchaturvedi reviewed (changes requested):
Requested few more changes

mgaurav99 committed:
Corrected  documentation as per reviews

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,151 @@
+    /// <summary>
+    /// WeatherReportServiceTests
+    /// Contains unit tests for the class,
+    /// validating its behavior under various temperature conditions and input scenarios.
```
avoid using ing

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,151 @@
+        readonly Mock<ITemperatureService> _mockTemperatureService;
+        readonly Mock<IDateTimeService> _mockDateTimeService;
+        readonly IWeatherReportService _weatherService;
+        DateTime sampleDate = DateTime.Parse("2023-10-01 12:00:00");
```
make sampleDate readonly and use 
DateTime.Parse("2023-10-01 12:00:00", CultureInfo.InvariantCulture)

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/WeatherReportService.cs:
```diff
@@ -0,0 +1,68 @@
+        public string GetClimate(string city)
+        {
+            DateTime currentDateTime = _dateTimeService.GetCurrentDateTime();
+            _loggerService.LogInfo($"This function called for city:{city} at {currentDateTime} ");
```
remove end space in all LogInfo

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,151 @@
+            string msg = _weatherService.GetClimate(city);
+
+            //Assert
+            _mockLoggerService.Verify(x => x.LogInfo($"This function called for city:{city} at {sampleDate} "));
```
remove end space in all LogInfo

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,151 @@
+            //Assert
+            _mockLoggerService.Verify(x => x.LogInfo($"This function called for city:{city} at {sampleDate} "));
+            _mockTemperatureService.Verify(x => x.GetTemperature());
+            Assert.Equal(expectedValue, msg);
```
Suggestion:
Use FluentAssertion

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/WeatherReportService.cs:
```diff
@@ -0,0 +1,68 @@
+            if (String.IsNullOrWhiteSpace(city))
+            {
+                _loggerService.LogWarning("No city provided!");
+                throw new ArgumentNullException("Provide city!");
```
keep all contants in Constants.cs

mgaurav99 replied:
these values are not used for application related tasks or application level events  but for general tests. Putting them in separate file would make code readability a bit tough as i need to look into another file everytime to understand the code.

vikchaturvedi commented on WeatherxUnit/WeatherReport/WeatherReportServiceTests.cs:
```diff
@@ -0,0 +1,151 @@
+        readonly Mock<ILoggerService> _mockLoggerService;
+        readonly Mock<ITemperatureService> _mockTemperatureService;
+        readonly Mock<IDateTimeService> _mockDateTimeService;
+        readonly IWeatherReportService _weatherService;
```
change object name to _weatherReportService

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/WeatherReportService.cs:
```diff
@@ -0,0 +1,68 @@
+            {
+                climate = "warm";
+            }
+
```
extra return

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/ConsoleLoggerService.cs:
```diff
@@ -0,0 +1,44 @@
+﻿
```
extra return

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/ConsoleLoggerService.cs:
```diff
@@ -0,0 +1,44 @@
+        /// <summary>
+        /// Log errors to console using red as foreground color
+        /// </summary>
+        /// <param name="message">The error message to be loggged. </param>
```
remove space at the end of the param name documentation

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/DateTimeService.cs:
```diff
@@ -0,0 +1,22 @@
+﻿
```
extra return

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/WeatherReportService.cs:
```diff
@@ -0,0 +1,68 @@
+        /// <exception cref="ArgumentNullException">
+        /// /// Thrown if <paramref name="logger"/>, <paramref name="temperature"/>, or <paramref name="dateTime"/> is <c>null</c>.
+        /// </exception>
+        public WeatherReportService(ILoggerService logger, ITemperatureService temperature, IDateTimeService dateTime)
```
indent parameters in next line

vikchaturvedi commented on WeatherxUnit/WeatherxUnit/Classes/WeatherReportService.cs:
```diff
@@ -0,0 +1,68 @@
+        /// </exception>
+        public WeatherReportService(ILoggerService logger, ITemperatureService temperature, IDateTimeService dateTime)
+        {
+            _loggerService = logger ?? throw new ArgumentNullException(nameof(logger));
```
unneccessary throwing exception

mgaurav99 committed:
Created test case for three other services:
ConsoleLoggerService
DateTimeService
TemperatureService

Added few more test cases for WeatherReportService
Used fluent assertions.
Included code coverage reports.
